src/onnx/parse_gridsample.cpp

```diff
@@ -0,0 +1,432 @@
+                        const instruction_ref& h,
+                        const instruction_ref& w,
+                        size_t n,
+                        size_t c,
+                        std::vector<instruction_ref>& indices,
+                        std::vector<instruction_ref>& validation,
+                        bool validate = true) const
+    {
+        static auto nc_shape = migraphx::shape{m_grid->get_shape().type(), {2}};
+        auto nc              = info.add_literal(migraphx::literal{nc_shape, {n, c}});
+        auto w_clamp = validate ? info.add_common_op("clip", w, m_zero_l, m_width_max_l) : w;
+        auto h_clamp = validate ? info.add_common_op("clip", h, m_zero_l, m_height_max_l) : h;
+        auto nchw    = info.add_instruction(make_op("concat", {{"axis", 0}}), nc, h_clamp, w_clamp);
+        indices.push_back(nchw);
+        if(validate)
+        {
+            auto h_valid = info.add_common_op("equal", h, h_clamp);
+            auto w_valid = info.add_common_op("equal", w, w_clamp);
+            auto valid   = info.add_common_op("logical_and", h_valid, w_valid);
+            validation.push_back(valid);
+        }
+    }
+
+    static instruction_ref concat_on_first_dim(const onnx_parser::node_info& info,
+                                               std::vector<instruction_ref> instructions)
+    {
+        if(instructions.empty())
+            MIGRAPHX_THROW("PARSE_GRID_SAMPLE: can't concatenate empty list of instructions");
+
+        auto ret = instructions.at(0);
+        std::for_each(
+            std::next(instructions.begin()), instructions.end(), [&info, &ret](auto& ins) {
+                ret = info.add_instruction(make_op("concat", {{"axis", 0}}), ret, ins);
+            });
+        return ret;
+    }
+
+    inline bool has_border_padding() const { return m_padding == "border"; }
+
+    virtual instruction_ref sample(const onnx_parser::node_info& info) = 0;
+};
+
+struct nearest_sampler : grid_sampler
+{
+    instruction_ref m_round_x;
+    instruction_ref m_round_y;
+
+    nearest_sampler(const instruction_ref& input,
+                    const instruction_ref& grid,
+                    bool align,
+                    std::string&& padding)
+        : grid_sampler(input, grid, align, std::move(padding))
+    {
+    }
+
+    void setup(const onnx_parser::node_info& info) override
+    {
+        grid_sampler::setup(info);
+        m_round_x = info.add_common_op("nearbyint", m_unnorm_x);
+        m_round_y = info.add_common_op("nearbyint", m_unnorm_y);
+    }
+
+    instruction_ref sample(const onnx_parser::node_info& info) override
+    {
+        setup(info);
+        std::vector<instruction_ref> indices;
+        std::vector<instruction_ref> validation;
+        static auto nhw_shape = migraphx::shape{migraphx::shape::int64_type, {3}};
+        bool validate         = not has_border_padding();
+        for(size_t n = 0; n < m_batch; n++)
```

Comment:
Get these to run for pre increment here for these for loops since worst case they're  O(n^4) (N,C, H, W)

Reply:
Maybe I don't get it right, but are you hinting that we should iterate over touples of possible combinations instead of the nested for loops?

Reply:
Start at -1 and change your loops to perform ++n vs n++
 
`( size_t n=-1; n < m_batch; ++n)`

When you n++ you create a temp variable so nesting this you're doing a bunch of extra create/copy/destroy ops on the background.

From in notes: https://en.cppreference.com/w/cpp/language/operator_incdec

_Because a temporary copy of the object is constructed during post-increment and post-decrement, pre-increment or pre-decrement operators are usually more efficient in contexts where the returned value is not used._

Reply:
> When you n++ you create a temp variable so nesting this you're doing a bunch of extra create/copy/destroy ops on the background.

For trivial types this shouldn't matter, the compiler will always optimize out the unused temporary.